### ISSUE TITLE
Feature: Last status timestamp metrics

### ISSUE THREAD

**Issue by travisgroth** [epoch8/airflow-exporter#43]
Using the current metrics layout, it becomes challenging to create alerts if you have a scenario where a DAG could not succeed for several periods.  The failed gauge will always read non-zero if there are historical gaps, even if subsequent runs may have succeeded.  In our case, these gaps might be acceptable.  Here's a DAG history that illustrates this:

```
State	Dag Id	                         Execution Date	Run Id	
success	mydag-daily_with-gap-in-success	01-15T00:00:00	scheduled__2019-01-15T00:00:00	
success	mydag-daily_with-gap-in-success	01-14T00:00:00	scheduled__2019-01-14T00:00:00	
success	mydag-daily_with-gap-in-success	01-13T00:00:00	scheduled__2019-01-13T00:00:00	
success	mydag-daily_with-gap-in-success	01-12T00:00:00	scheduled__2019-01-12T00:00:00	
success	mydag-daily_with-gap-in-success	01-11T00:00:00	scheduled__2019-01-11T00:00:00	
success	mydag-daily_with-gap-in-success	01-10T00:00:00	scheduled__2019-01-10T00:00:00	
success	mydag-daily_with-gap-in-success	01-09T00:00:00	scheduled__2019-01-09T00:00:00	
success	mydag-daily_with-gap-in-success	01-08T00:00:00	scheduled__2019-01-08T00:00:00	
success	mydag-daily_with-gap-in-success	01-07T00:00:00	scheduled__2019-01-07T00:00:00	
success	mydag-daily_with-gap-in-success	01-06T00:00:00	scheduled__2019-01-06T00:00:00	
success	mydag-daily_with-gap-in-success	01-05T00:00:00	scheduled__2019-01-05T00:00:00	
success	mydag-daily_with-gap-in-success	01-01T00:00:00	scheduled__2019-01-01T00:00:00	
success	mydag-daily_with-gap-in-success	2018-12-31T00:00:00	scheduled__2018-12-31T00:00:00	
success	mydag-daily_with-gap-in-success	2018-12-30T00:00:00	scheduled__2018-12-30T00:00:00	
success	mydag-daily_with-gap-in-success	2018-12-29T00:00:00
```

The resulting metrics are as follows:

```
airflow_dag_status{dag_id="mydag-daily_with-gap-in-success",owner="airflow",status="failed"} 15.0
airflow_dag_status{dag_id="mydag-daily_with-gap-in-success",owner="airflow",status="running"} 0.0
airflow_dag_status{dag_id="mydag-daily_with-gap-in-success",owner="airflow",status="success"} 365.0
```

(a) I'm open to suggestions on how to handle this type of situation correctly.  A simplistic check for failed count will alert once and never again.  Checking for rate feels a bit fragile.

(b) a pattern that is common in batch monitoring is to also expose the last time a job was in X state, so that you can gauge if you've succeeded since you've failed and how long it has been since you've succeeded (failed or not).  This can catch a variety of scenarios including airflow being stuck for whatever reason.  I'm fairly certain exposing this as a metric would also help the concrete scenario I've proposed here.

Is this possible?  A suggestion might be

airflow_dag_status_time{dag_id="mydag-daily_with-gap-in-success",owner="airflow",status="success"} 1547639402
airflow_dag_status_time{dag_id="mydag-daily_with-gap-in-success",owner="airflow",status="failed"} 1547639102

This could be extended for tasks too though I'm not sure that is useful.  

Open to feedback, of course.  Thanks.

**Comment by elephantum** [comment]
Will it be useful to have explicit “last_status” metric?

**Comment by travisgroth** [comment]
WRT to manually setting success - yes that's a reasonable position.  In our case we have external interactions that we don't want engineers to have to intervene in until there are excessive failure periods (say 3 failed runs in a row or 3 days of failure).  Manually marking them successful will prove challenging to scale for our use case.

Structure wise - I'm quite interested in getting the timestamps as a value, as I think it provides a lot of derivative utility.  Examples - historical runtime, stuck scheduler, read-only database, various other hard to predict conditions where a job doesn't ever "fail" but also didn't succeed.  Just a last status doesn't capture those properly.

Providing concrete time stamps also keeps scrape and job intervals from being intertwined.  This also makes it pushgateway friendly and resilient to missed or long scrape periods.

An additional basic status field could be useful for simplified queries, though.  It would save having to derive the last state yourself if you are interested in a red/green dashboard.  I don't think it is quite enough for the use cases I have in mind though.  Thoughts?

**Comment by elephantum** [comment]
Fixed by #96
